column types and description

pivnicek opened the pull request:
Fixes #3038

Proposed changes:

- Add description and dataType form/info for columns in table detail view

---


Up to date screenshot:

![Screenshot 2019-04-09 at 15 51 17](https://user-images.githubusercontent.com/419849/55805752-5a421d80-5adf-11e9-92c6-e682519224b0.png)

natocTo commented:
I see this is very early stage. So maybe I will wait few more days than I try bigger review.
But I am here and ready for any help if you need.

pivnicek commented:
Hi @natocTo ,  I have it finally "basically" working, there are still some minor issues to fix but I think it may be a good time to discuss doing it the best way, especially WRT https://github.com/keboola/kbc-ui/pull/2762.

Could you please have a look and let me know what you think?  Don't worry about minor issues, I'm just a little worried that I may have done bigger things stupid.  Thanks man.

natocTo commented:
Hi, yes definitely. I start with this tomorrow morning. Or maybe even tonight, I will see.

natocTo commented on src/scripts/modules/storage-explorer/react/pages/Table/TableColumn.jsx:
```diff
@@ -50,36 +93,60 @@ export default React.createClass({
+        <ColumnDetails
+          columnId={this.getColumnId(column)}
+          columnName={column}
+          machineDataType={this.getDataType(this.props.machineColumnMetadata.get(column))}
```
`machineColumnMetadata.get(column)` get return undefined, so `getDataType` should can considered it, or here can be something like `this.props.machineColumnMetadata.get(column, Map())`

natocTo commented on src/scripts/modules/storage-explorer/react/pages/Table/TableColumn.jsx:
```diff
@@ -50,36 +93,60 @@ export default React.createClass({
+          columnId={this.getColumnId(column)}
+          columnName={column}
+          machineDataType={this.getDataType(this.props.machineColumnMetadata.get(column))}
+          userDataType={this.getDataType(this.props.userColumnMetadata.get(column))}
```
Same as one line above.

natocTo commented on src/scripts/modules/storage-explorer/react/pages/Table/ColumnDetails.jsx:
```diff
@@ -0,0 +1,142 @@
+        "Unknown"
+      )
+    }
+  },
```
I think I would use at least <span /> here. I am not sure if return string is supported in React 14 but maybe yes.

natocTo commented on src/scripts/modules/storage-explorer/react/pages/Table/ColumnDetails.jsx:
```diff
@@ -0,0 +1,142 @@
+  renderTypeForm() {
+    return (
+      <Table>
+        <Row>
```
`Table -> Row` is not allowed, it render `<table><div>` which is not supported.
You can remove `Table` component or use `Grid` component instead.

natocTo commented on src/scripts/modules/storage-explorer/react/pages/Table/ColumnDetails.jsx:
```diff
@@ -0,0 +1,142 @@
+      'KBC.datatype.length': this.state.userDataType.get('length'),
+      'KBC.datatype.nullable': this.state.userDataType.get('nullable'),
+    }));
+  }
```
When I try save column metadata I got error: "Invalid structure. Metadata must be provided as an array of objects with 'key' and 'value' members". So this should be updated to right format.

pivnicek replied:
oh, strange, I hadn't got that error yet!  Dik!

pivnicek commented:
Thanks @natocTo , I'll update!

pivnicek commented on src/scripts/modules/storage-explorer/react/pages/Table/TableColumn.jsx:
```diff
@@ -28,6 +38,39 @@ export default React.createClass({
+
+    const nativeType = metadata.find((entry) => {
+      return entry.get('key') === 'KBC.datatype.type';
+    }, null, Map());
```
ah, but if baseType doesn't exist then I want to return empty Map.  That is why it is different.

natocTo replied:
Yes you are right :) sorry about this

pivnicek commented:
ok @natocTo I've covered your first feedback.  This afternoon I will touch-up the aesthetics and update the get  datatype method for transformations/writers to use the user provided ones if found.

natocTo commented:
I forgot to mention that instead of `React.PropTypes` is now used `prop-types` library, and instead of `React.createClass` is now used `create-react-class` library.

pivnicek commented:
Yes, I noticed that when I resolved conflicts.  ✅ 
I need to add it to my new file too.  diky

pivnicek commented:
@natocTo , I've made it a little prettier,  could you please have a look?  Also, I'm not sure what to do with the Save button (I don't know which common component to use).    Thanks man.

natocTo commented on src/scripts/modules/storage-explorer/react/pages/Table/ColumnDetails.jsx:
```diff
@@ -0,0 +1,170 @@
+      'KBC.datatype.basetype': this.state.userDataType.get('baseType'),
+      'KBC.datatype.length': this.state.userDataType.get('length'),
+      'KBC.datatype.nullable': this.state.userDataType.get('nullable', false),
+    }));
```
I think it is good to add some "loading" state to signalize user that you saving the metadata, you can also use this new state to disable save button during saving.

```js
this.setState({ isSaving: true });
saveColumnMetadata(
  this.props.columnId,
  Map({
    'KBC.datatype.basetype': this.state.userDataType.get('baseType'),
    'KBC.datatype.length': this.state.userDataType.get('length'),
    'KBC.datatype.nullable': this.state.userDataType.get('nullable', false)
  })
).finally(() => {
  this.setState({ isSaving: false });
});
```

If you want to use this you must also return a Promise from `saveColumnMetadata` action.

natocTo commented:
Some more thoughts.
I did not check the UI so much, I will check more tonight.

pivnicek commented on src/scripts/modules/storage-explorer/react/pages/Table/ColumnDetails.jsx:
```diff
@@ -0,0 +1,170 @@
+                value={this.state.userDataType.get('baseType')}
+                options={this.baseTypeOptions()}
+                onChange={this.handleBaseTypeChange}
+              />
```
good point, it should be clearable, and saving it as empty should delete the user defined type.  Thank you!

natocTo commented:
> ... and update the get datatype method for transformations/writers to use the user provided ones if found.

I thinking about this. Maybe this could be in new PR after this one. What do you think?

natocTo commented on src/scripts/modules/components/stores/MetadataStore.js:
```diff
@@ -122,6 +128,15 @@ var MetadataStore = StoreUtils.createStore({
+  },
+
+  getLastUpdatedByColumnMetadata: function(tableId) {
+    const lastUpdateInfo = this.getTableLastUpdatedInfo(tableId);
```
One more :) `getTableLastUpdatedInfo` can return null.

pivnicek replied:
done

natocTo commented on src/scripts/modules/storage-explorer/react/pages/Table/ColumnDetails.jsx:
```diff
@@ -0,0 +1,170 @@
+  handleSaveDataType() {
+    saveColumnMetadata(this.props.columnId, Map({
+      'KBC.datatype.basetype': this.state.userDataType.get('baseType'),
+      'KBC.datatype.length': this.state.userDataType.get('length'),
```
The `this.state.userDataType.get('length')` can be undefined when you select type like "Date". 
Maybe `this.state.userDataType.get('length', null)` would be better. Same for `this.state.userDataType.get('baseType')` if is possible to send null value. It is also undefined by default.

Or make save button enabled only for valid state, whatever it is.

pivnicek replied:
I think I'll make it delete/clear the user-defined type if they save it with no baseType.  Will put a help block there.

pivnicek commented:
Hi @natocTo , I've made quite a few changes this morning, mainly is using constants and adding metadata delete, but I think I've resolved everything.

natocTo commented:
I see lots of new commits :) I do review after lunch.

natocTo commented:
I am not sure about actual labels, but maybe we can try do it more compact. Something like this.
![column](https://user-images.githubusercontent.com/12331181/55324598-e078b700-5482-11e9-9f19-ad85e3f905fb.png)

Also if no System value is available I see no text.
![blank](https://user-images.githubusercontent.com/12331181/55324959-e0c58200-5483-11e9-962a-10525963c134.png)

Ou, it is a bug because `machineDataType` now return empty Map, so it can be updated to something like this:

```jsx
enderSystemValue() {
  if (!this.props.machineDataType.count()) {
    return <FormControl.Static>Unknown</FormControl.Static>;
  }
  
  return (
    <div>
      {this.props.machineDataType.get(DataTypeKeys.BASE_TYPE)}
      {
        this.props.machineDataType.get(DataTypeKeys.LENGTH)
          ? '(' + this.props.machineDataType.get(DataTypeKeys.LENGTH) + ')'
          : ''
      }
      <br/>
      <Label bsStyle="info">{this.props.machineDataType.get('provider')}</Label>
    </div>
  );
}
```

I use `FormControl.Static` instead of `span` for better vertical alingment.

natocTo commented on src/scripts/modules/storage-explorer/react/pages/Table/ColumnDetails.jsx:
```diff
@@ -0,0 +1,198 @@
+          <FormGroup>
+            <Col componentClass={ControlLabel} xs={4}>&nbsp;</Col>
+            <Col xs={8}>
+              <Button bsStyle="success" onClick={this.handleSaveDataType}>
```
button should be disabled while `isSaving` is true

natocTo replied:
When type with length is chosen, the length should be entered otherwise button should be disabled as well. Or not?

pivnicek replied:
Length is optional even for ones that allow it.

natocTo replied:
Try update some Date column to String for example, and leave length without value. I got error because sometimes data looks like

```js
{
    KBC.datatype.basetype: "STRING"
    KBC.datatype.length: undefined
    KBC.datatype.nullable: false
}
```
![error](https://user-images.githubusercontent.com/12331181/55331609-edea6d00-5493-11e9-822b-7e3aebc7f647.png)


Here you can see how data look like:

![data](https://user-images.githubusercontent.com/12331181/55331667-08bce180-5494-11e9-8020-6670e40e3af3.png)

natocTo replied:
Also when I change type from Number(10) do Date, the data look like this.

```js
{
    KBC.datatype.basetype: "DATE"
    KBC.datatype.length: "10"
    KBC.datatype.nullable: false
}
```

The `length` should not be here I think.

pivnicek replied:
good catch, thank you!

pivnicek replied:
ok, should be fixed, thanks again.

natocTo commented on src/scripts/modules/storage-explorer/react/pages/Table/TableColumn.jsx:
```diff
@@ -30,6 +42,54 @@ export default createReactClass({
+    return this.props.table.get('id') + '.' + columnName;
+  },
+
+  deleteUserType(columnName) {
```
If I select some type, and than I try to remove it, it does nothing (I see no request). It is working only after reload.

pivnicek replied:
confirmed :-/

pivnicek replied:
ah!, it's because when you save a type, the `userDefinedType` property isn't updated

pivnicek replied:
this should be fixed now.

pivnicek commented:
@natocTo I was thinking to not show that `unknown` at all and instead put there sample values (users would find that helpful I think.  I'm trying to import sampleColumnValues as a prop but I see that it's defined inside the component https://github.com/keboola/kbc-ui/blob/master/src/scripts/modules/storage-explorer/react/components/DataSample.jsx.  So I'm not sure about the best way to do it.  But it'll probably need a refactor I guess.

natocTo commented:
I would prefer link or hint about info that there is also "Data sample" tab. Which is entirely dedicated to show sample values.

There is also `ColumnDataPreview` component. But I am not sure if can be used here.

pivnicek commented:
Yes, anyhow, in the new storage design columns have their own tab so it can be solved there/then I guess.

pivnicek commented:
Alright @natocTo I think all the issues are resolved.

natocTo commented:
Ok, I do one more review later today. Good work.

natocTo commented on src/scripts/modules/storage-explorer/react/pages/Table/ColumnDetails.jsx:
```diff
@@ -0,0 +1,213 @@
+            {
+              this.renderSystemValue()
+            }
+          </FormGroup>
```
There is two `FormGroup` in row. Here first one and second one at `this.renderSystemValue()`. Should be used only one to have same alignment with other part of form.

natocTo commented on src/scripts/modules/storage-explorer/react/pages/Table/ColumnDetails.jsx:
```diff
@@ -0,0 +1,213 @@
+    return {
+      userDataType: props.userDataType,
+    }
+  },
```
This function is never called I think. `getStateFromStores` is called by `createStoreMixin` and is used together with some stores. But there is no direct pinned store.

natocTo replied:
If you need update `userData Type` you can use `componentDidUpdate` React lifecycle hook.

natocTo replied:
Example:

```jsx
componentDidUpdate(prevProps) {
    if (!prevProps.userDataType.equals(this.props.userDataType)) {
      this.setState({ userDataType: this.props.userDataType });
    }
  }
```

natocTo commented on src/scripts/modules/storage-explorer/react/pages/Table/ColumnDetails.jsx:
```diff
@@ -0,0 +1,213 @@
+    if (this.state.userDataType.get(DataTypeKeys.BASE_TYPE)) {
+      saveColumnMetadata(
+        this.props.columnId,
+        Map(this.state.userDataType)
```
Another small thing. `this.state.userDataType` is already Map I think.

pivnicek replied:
Yes, it is.  It is strange to me that this doesn't break it :-/

natocTo commented on src/scripts/modules/storage-explorer/react/pages/Table/ColumnDetails.jsx:
```diff
@@ -0,0 +1,213 @@
+          <FormGroup>
+            <Col componentClass={ControlLabel} xs={4}>&nbsp;</Col>
+            <Col xs={8}>
+              <Button bsStyle="success" onClick={this.handleSaveDataType} disabled={this.state.isSaving}>
```
I would add function `disabled()` where you can also check if data is changed, and prevent save if not.

Example:

```jsx
isDisabled() {
  return this.state.isSaving || this.props.userDataType.equals(this.state.userDataType);
}
```

natocTo replied:
But is bad when I delete current type, the object look like `{}` and props return default `{KBC.datatype.basetype: "INTEGER", KBC.datatype.length: null, KBC.datatype.nullable: false}` so Save button is enabled anyway. Maybe you figure it out.

pivnicek replied:
props are returning like that because they're not updating properly I think.  After the delete the prop and state should both be `{}`  .  I hope componentDidMount will solve that.

natocTo commented on src/scripts/modules/storage-explorer/react/pages/Table/TableDetail.jsx:
```diff
@@ -30,7 +32,7 @@ import LatestImports from './LatestImports';
 
 export default createReactClass({
-  mixins: [createStoreMixin(TablesStore, BucketsStore, ApplicationStore, FilesStore)],
+  mixins: [createStoreMixin(TablesStore, BucketsStore, ApplicationStore, FilesStore, ColumnsLocalStore)],
```
The `MetadataStore` missing here. To be able to reflect changes in store after successful request.

pivnicek commented:
Thanks so much, I'm on it.

natocTo commented on src/scripts/modules/components/stores/MetadataStore.js:
```diff
@@ -32,7 +32,13 @@ var MetadataStore = StoreUtils.createStore({
 
   getTableColumnsMetadata: function(tableId) {
-    return _store.getIn(['metadata', 'tableColumns', tableId]);
+    return _store.getIn(['metadata', 'tableColumns', tableId], Map());
```
The table metadata is here loadad from `['metadata', 'tableColumns', tableId]` path.
But after successfull update request, action is fired (METADATA_SAVE_SUCCESS) it match this case:

```jsx
case ActionTypes.METADATA_SAVE_SUCCESS:
    _store = _store.setIn(['metadata', action.objectType, action.objectId], fromJS(action.metadata));
    _store = _store.deleteIn(['savingMetadata', action.objectType, action.objectId, action.metadataKey]);
    return MetadataStore.emitChange();
```

If you log `action` object, you can see that `objectType` is not `tableColumns` but only `column`. So data is saved to another place. Than the TableDetail component try read new values from store, but he found only the old ones. Or I am wrong?

Also I thought we will use metadata store only for editing, and we will read and update metadata direct on table object for example. But the PR which was sent is not still merged so it can be updated later.

pivnicek replied:
> Or I am wrong?
doh! you are not wrong!

pivnicek commented:
I've got it all except for the last one, but I've an idea how to solve it.  will take care of it right after lunch.  Going into town now.

pivnicek commented:
ok @natocTo, the last commit changes it to get the data from the `[metadata, column, columnId]` instead.  It’s a hack, but we can refactor it when we merge #2762 imho.

natocTo commented:
Can I sent few commits? But still I am not sure about the last changes.

pivnicek commented:
@natocTo yes please!

pivnicek commented on src/scripts/modules/storage-explorer/react/pages/Table/ColumnDetails.jsx:
```diff
@@ -78,8 +76,7 @@ export default createReactClass({
           <FormGroup>
-            <Col componentClass={ControlLabel} xs={4}>&nbsp;</Col>
-            <Col xs={8}>
+            <Col xs={8} xsOffset={4}>
```
👍

natocTo commented:
I will look more deep on MetaStore, but later today.

pivnicek commented on src/scripts/modules/storage-explorer/react/pages/Table/ColumnDetails.jsx:
```diff
@@ -191,6 +187,10 @@ export default createReactClass({
 
+  isDisabled() {
+    return this.state.isSaving || this.props.userDataType.equals(this.state.userDataType);
+  },
```
thank you! sorry I forgot about that in my rush this morning:)

pivnicek commented:
@natocTo ok, thanks for the commits, looks good.

Regarding the metadata store, I'd like to fix it in https://github.com/keboola/kbc-ui/pull/2762.

natocTo commented:
I did some changes, I think now it kind of works, but I am not sure how happy will be @ujovlado :))) the Metadata store need refactor as soon as possible :)

pivnicek commented:
@natocTo I thought of doing it your way:), looking at it now it's nicer.  But yes, I agree, we need to refactor that asap.  Thanks so much for your help.  I'll request review from @ujovlado .

ujovlado commented:
The `length` input field is a bit buggy (switching between controlled/uncontrolled) because its value is undefined.

Is it ok to have structure like this?

```js
{"KBC.datatype.basetype" => "STRING"}
{"KBC.datatype.length" => undefined}
{"KBC.datatype.nullable" => false}
```
I would say no. I mean, what is the point of having value `undefined`?

It is caused by `getDataType` fn in `TableColumn.jsx`. That fn should be extracted somewhere and tests should be added. So we'll see what it is doing.

ujovlado commented:
I'll add here small cosmetic changes.

ujovlado commented:
And few more things:

- we try to not use single character variables like `k` or `v`
- Why is there `return this.setState(...)` in few places instead of `if ... else if ... else`? It is very hard to understand, that you are using it stop execution ...

ujovlado commented:
This is WTF for me:

```js
if (cond1) {
  return this.setState(...);
}
if (cond2) {
  return this.setState(...);
}
return this.setState(...);
```
And it is very hard to understand you meant this:
```js
if (cond1) {
 this.setState(...);
}
else if (cond2) {
 this.setState(...);
}
else {
 this.setState(...);
}
```

pivnicek commented:
Thanks @ujovlado !! I'll update right away

pivnicek commented:
@ujovlado I fixed the minor issues and I think the length issue is solved by the refacored `getDataType` method that I moved to utils and made tests for.  btw, I like the `tdd` test watch thing, very cool.

pivnicek commented:
shit, didn't realise the invalid base type test was still failing.  fixed now.

ujovlado commented:
@pivnicek what is the expected behaviour for length deletion? Empty string should be saved or whole `KBC.datatype.length` item deleted?

pivnicek commented:
Either is ok imo.  But when the whole type is deleted then the length gets deleted too.

pivnicek commented:
But if old type was `String(32)` and it's changed to date, length should be deleted.

pivnicek commented:
Shit, just checked that and it keeps that 32 :(.  If saved type does not support length, it should deleted the length data that was previously there imo.  Damn, I'll update it now.

ujovlado commented:
I'm only asking because if I have `KBC.datatype.length` set to `"123"` than aftere deletion there will be `KBC.datatype.length` set to `""` (empty string). Imho we should delete it that case? What do you think?

pivnicek commented:
I think having it that way (length = "") is fine for types that support length.  But I found a related different bug -- if old type was String(123), and you switch to Date and save, the type stored will be `type = Date` & `length=123` which is a problem.

pivnicek commented:
But, to solve both, it may be easiest to do it as you suggest, would solve both issues.

pivnicek commented:
@ujovlado I think my last commit should solve it (I checked and think it's ok), but I'm not sure how to best handle it tbh

pivnicek commented:
ok, @ujovlado I refactored it a bit and it seems to work as I expect, but please tell me if it is debyl, because that is definitely possible

ujovlado commented:
Works fine, I'm not sure about that deletion - why do you set it to empty string `""` and then delete it if it's set to empty string? Shouldn't we delete it directly if length is not supported? @natocTo what do you think?

ujovlado commented:
And last thing I'm missing there is to see what is set, so I don't have to open each panel just to check it.

![Screenshot 2019-04-05 at 13 17 25](https://user-images.githubusercontent.com/419849/55624093-2d230180-57a5-11e9-93bd-94cc63a403f7.png)

natocTo commented:
We talked about length quite a lot. I will pull latest changes and check it how it look now.

ujovlado commented:
^ I was talking about this commit https://github.com/keboola/kbc-ui/pull/3102/commits/24608e2fe5890547b53dd8b3cf748146357bfcc1

natocTo commented:
Question. The "system" metadata has only type defined or can have also length?

pivnicek commented:
> The "system" metadata can has only type defined or can have also length?

System metadata is fully defined.

> And last thing I'm missing there is to see what is set, so I don't have to open each panel just to check it.

@ujovlado That looks great!  We should show there label for provider for type if defined by system too, so you know if it is system defined or user defined.

natocTo commented:
I think better and more reliable to the future will be compare `userDataType` and `this.props.userColumnMetadata` in `saveUserType`.

Everything what is not in "new" userDataType should be deleted. Does it makes sense?
But if we will never have any more metadata like `length` actual code works too.

pivnicek commented:
Great idea @natocTo !  I will do it that way.

@ujovlado can you please push your version with the Type displayed  in the header?

ujovlado commented:
I can, but which metadata should be displayed there? user only? Looks like I am missing something.

natocTo commented:
What happen when System metadata will be `STRING(10)` and we change to `NUMBER` without length? I am just asking if we do not use NUMBER from user metadata and lenght (10) from system metadata. 

That is why I thought that lenght should be mandatory when is selected type which support length. I am just curious.

ujovlado commented:
I wonder if something like this wouldn't be better: https://github.com/keboola/kbc-ui/issues/3038#issuecomment-473833568

pivnicek commented:
Order should be if (usertype) displayUserType else if (systemType) displaySystemType else ""

Like this: https://github.com/keboola/kbc-ui/pull/3163/commits/3c6ee3926e56d7980002261461793646808b0f08

That ^ is to use types in IM

ujovlado commented:
ok, understand, so if base type for provider is there, the whole "set" will be used from provider.

pivnicek commented:
Exactly

pivnicek commented:
If I get the energy I'm going to attempt the MetadataStore refactor over the weekend, to clean all this up and centralise these methods.

ujovlado commented:
There's already a PR for that, I'll check it after this PR is merged. We can discuss it later.

pivnicek commented:
> There's already a PR for that, I'll check it after this PR is merged. We can discuss it later.

I know, but it is outdated and I know how I want to do it now :-).  Anyways it would only be an exercise.
Next priority for this is #3163 which makes these types useful.

pivnicek commented:
😢  https://github.com/keboola/internal/issues/136#issuecomment-480253953

pivnicek commented:
@ujovlado if you push your commit I can update the header as described.

ujovlado commented:
Wait a sec.

ujovlado commented:
@pivnicek it is there ... it is hard coded to show user data types ... feel free to update it.

pivnicek commented:
Thanks! I will, mmt

pivnicek commented:
@ujovlado how about this:
![image](https://user-images.githubusercontent.com/569139/55630855-f8b84100-57b6-11e9-87a7-31d0382dee1b.png)

ujovlado commented:
I would prefer to render it without additional `<code>` tag.

pivnicek commented:
ok, sure!

ujovlado commented:
Anyway, I found an issue. When deleting data type.

Reproduce:

- define user data type e.g. `string(111), Nullable`
- click on "x" in Type select
- click save  // button remains enabled
- click save (produces error)

^ and data type is still shown as `string(111), Nullable` ... after refresh I can see it is back on system data type

pivnicek commented:
Wait, it is not possible to delete system types

pivnicek commented:
But it shouldn't produce that error, I will fix it right now

pivnicek commented:
doing what @natocTo suggested will solve this.  mmt.

pivnicek commented:
It is a problem with the promise not returning on delete and the prop not updating.

pivnicek commented:
gah :-( `METADATA_DELETE_SUCCESS` wasn't hooked with the store.

ujovlado commented:
I was talking about user data types. Define it, and then try to remove it.

pivnicek commented:
Yeah, I know, delete action was broken, it wasn't propagated to store.  fixed now.

natocTo commented:
Taky ještě pošlu něco:

TableColumn
- tady se to Promise array dával třeba Map nebo List místo samotné promise, tak trochu upraveno
- navíc nekontroluji ty hodnotu `length = ""` ale jen zda existuje nebo ne, na ColumnDetail je mažu tak by to mělo být ok

ColumnDetails
- pidi úprava pár render funkcí, render `null` pokud se nesplní podmínka
- baseTypeOptions se převádělo na Map a pak hned na Array, tak jsme to nechal rovnou array jako je to definované
- handleLengthChange - pokud je length "" tak mažu danou věc jinak nastavuji
- handleBaseTypeChange - zjednodušení pro tu první podmínku, tam se dal objekt a pak se vše mazalo, tak tam rovnou posílám prázdný Map
- handleBaseTypeChange  - pokud není povolené nastavení velikosti tak mažu velikost místo nastavení ""
- handleSaveDataType - tady se to zjednodušilo a už nehlídám tu length

natocTo commented:
Vím že třeba ve DB-WR signalizuje červeně, pokud uživatel píše do `length` něco nesmyslného. Nevím zda něco takového taky přidat sem. Nebo to prostě nechat na na zodpovědnosti uživatele.

natocTo commented:
Zkusil jsem to přidat pouze tak, že pokud není `length` validní, je uložení zakázané.

Rozdělil jsem to na více commitů, tak dyštak jen revert pokud to tu nechceme. Commit `validate type si not case sensitive` jsem přidal jako třetí i když hned v první commitu jsem na to přidal test, ale ono by se to asi dyštak vracelo celé takže snad jedno.

ujovlado commented:
Prijemne vylepsenia, uz to tam poslem, nech sa toho zbavime. ;)
